### ISSUE TITLE
Memory leak after upgrade to 5.10.1

### ISSUE THREAD

**Issue by MedicMomcilo** [OpenNebula/one#4154]
**Description**
We ran 5.8 for long period and decided to upgrade.
After upgrade, RAM usage is really high (over 20GiB) with OOM kicking in.

We followed the guide for upgrade and faced no issues during.

**To Reproduce**
Upgrade the cluster.

**Expected behavior**
Similar RAM usage/requirement.

**Details**
 - Affected Component: oned
 - Hypervisor: KVM
 - Version: 5.10.1

**Additional context**
Frontend runs on a dedicated physical host with 32GiB of RAM and 4 cores.

## Progress Status
- [ ] Branch created 
- [ ] Code committed to development branch
- [ ] Testing - QA
- [ ] Documentation 
- [ ] Release notes - resolved issues, compatibility, known issues
- [ ] Code committed to upstream release/hotfix branches
- [ ] Documentation committed to upstream release/hotfix branches

**Comment by rsmontero** [comment]
We've uploaded a patch for a couple of leaks. However they are "minor" and we are checking for additional problems.

We assume that 5.8 behavior was ok, in terms of memory consumption?

**Comment by MedicMomcilo** [comment]
Yes, memory consumption was not an issue with 5.8.
This looks like a major leak as we ran out of 32GiB of RAM + 16GiB of SWAP.

It gets worse during the night where we have heavy ONE activity:
- nightly snapshots for all CEPH-based VMs (creating and removing snapshots)
- image backups for large number of VMs (shutting down, snapshot, booting,...)

All of those come with a ton of side-requests for VM states, disk type information, etc.

**Comment by rsmontero** [comment]
So we've been able to reproduce the issue. It seems that those ton of side-requests generate events at a pace much greater that the current event driver can handle. This, in turns makes oned buffers to increase and eventually OOM takes part in the party.

As those API calls are info (and currently we do not see a clear use case for them) we have disabled events on those API calls. We have uploaded a patch to master and one-5.10 branch. 

Thanks for your feedback and detailed comments. Closing the issue

PS: As a side effect we have fixed those minor leaks :)

**Comment by MedicMomcilo** [comment]
Hi Ruben,

Unfortunately this doesn't solve memory leak on our end.
```
root@one0:~# oned --version
Copyright 2002-2019, OpenNebula Project, OpenNebula Systems        

OpenNebula 5.10.2 (not known) is distributed and licensed for use under the terms of the
Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0).
root@one0:~# free -m
              total        used        free      shared  buff/cache   available
Mem:          31998       29589         386           1        2023        1960
Swap:         16383        1818       14565
root@one0:~# systemctl restart opennebula
root@one0:~# free -m
              total        used        free      shared  buff/cache   available
Mem:          31998       10206       19869           1        1922       21355
Swap:         16383         169       16214
root@one0:~# uptime
 09:18:36 up 18:31,  2 users,  load average: 4.60, 4.44, 4.70
root@one0:~#
```
Is there any information I can collect and provide, to help with this resolution?

Kind regards,
Momo.

**Comment by rsmontero** [comment]
can you send the process information of oned. Maybe memory is taken by ruby processes.

**Comment by MedicMomcilo** [comment]
```
root@one0:~# pidof oned
21432
root@one0:~# cat /proc/21432/status 
Name:	oned
Umask:	0022
State:	S (sleeping)
Tgid:	21432
Ngid:	0
Pid:	21432
PPid:	1
TracerPid:	0
Uid:	9869	9869	9869	9869
Gid:	9869	9869	9869	9869
FDSize:	256
Groups:	6 1010 9869 
NStgid:	21432
NSpid:	21432
NSpgid:	21432
NSsid:	21432
VmPeak:	14209524 kB
VmSize:	13459800 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	12581588 kB
VmRSS:	12133652 kB
RssAnon:	12120728 kB
RssFile:	   12924 kB
RssShmem:	       0 kB
VmData:	12936916 kB
VmStk:	     132 kB
VmExe:	    3700 kB
VmLib:	   54932 kB
VmPTE:	   25320 kB
VmSwap:	    1520 kB
HugetlbPages:	       0 kB
CoreDumping:	0
Threads:	26
SigQ:	1/127159
SigPnd:	0000000000000000
ShdPnd:	0000000000010000
SigBlk:	fffffffe7ffbbefd
SigIgn:	0000000000001000
SigCgt:	0000000180000000
CapInh:	0000000000000000
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	0000003fffffffff
CapAmb:	0000000000000000
NoNewPrivs:	0
Seccomp:	0
Speculation_Store_Bypass:	thread vulnerable
Cpus_allowed:	f
Cpus_allowed_list:	0-3
Mems_allowed:	00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:	0
voluntary_ctxt_switches:	446
nonvoluntary_ctxt_switches:	10
root@one0:~#
```

Please let me know if there is anything else I can provide.

**Comment by rsmontero** [comment]
Let's try this, could you send the out put of the following commands:

Bigger page sizes of process:

```
pmap `pgrep oned` | sort -k2 -nr | head
```

Strings of the biggest page (so we may have a clue of the affected component):
```
cat /proc/`pgrep oned`/maps  | sort -k5 -nr |head -1 | sed -n 's/^\([0-9a-f]*\)-\([0-9a-f]*\) .*$/\1 \2/p' | read start end ; sudo gdb --batch --pid `pgrep oned` --ex "dump memory /tmp/page 0x$start 0x$end" >/dev/null 2>&1 ; strings -n10 /tmp/page
```

Stack trace of gdb threads:
```
sudo gdb -q -ex 'thread apply all bt' -ex 'detach' -ex 'quit' `which oned` `pgrep oned` 
```

Hope I paste the commands right. Note I added sudo to attach to oned process...

THANKS

**Comment by MedicMomcilo** [comment]
I've attached output as text file as it is a bit long.

Also, note that second command doesn't provide useful output, so I've also executed it without last part.

[output.txt](https://github.com/OpenNebula/one/files/4235259/output.txt)

**Comment by rsmontero** [comment]
It seems that in the second command you are not getting the right addresses. Could you debug the oneliner. after the sed, you should have start and end virtual address of the page. and use that for the dump command with 0x... and 0x... Gdb complains about invalid 0x address.

**Comment by rsmontero** [comment]
Also the debugging symbols are not present, did you compile it without -g option?

**Comment by christian7007** [comment]
Hello @MedicMomcilo,

Could you also execute this script at the frontend node: https://gist.github.com/christian7007/53e95e5e91acc09ed0a3e1d7abc37c8a

and execute  `onevm list` and `onezone list` while the script is running and send us the output so we can make sure that the fix we introduced is working for you and discard that problem?

You should see something similar to this:
```
$ ruby event-client.rb 
EVENT API one.zone.raftstatus 1
```

**Comment by MedicMomcilo** [comment]
Hi @rsmontero ,

I've deconstructed one-liner and managed to get the output:
```
root@one0:~# gdb --batch --pid `pgrep oned` --ex "dump memory /tmp/page 0x7fc84d567000 0x7fc84d568000" >/dev/null 2>&1 ; strings -n10 /tmp/page
.eh_frame_hdr
.init_array
.fini_array
.gnu_debuglink
root@one0:~#
```

Also, I've installed opennebula-dbgsym_5.10.2-1_amd64.deb which I suppose provides debugging symbols.

Hi @christian7007 ,

Unfortunatelly, script doesn't find necessary gem:

```
root@one0:~# /var/lib/one/event-client.rb 
Traceback (most recent call last):
	2: from /var/lib/one/event-client.rb:3:in `<main>'
	1: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
/usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require': cannot load such file -- ffi-rzmq (LoadError)
root@one0:~#
```

Please let me know if there is anything else I can do.

Kind regards,
Momo.

**Comment by christian7007** [comment]
Hello @MedicMomcilo,

I've updated the script: https://gist.github.com/christian7007/53e95e5e91acc09ed0a3e1d7abc37c8a

Could you try again?

**Comment by MedicMomcilo** [comment]
Hi @christian7007 

Indeed the script runs and does provide output as expected:
```
root@one0:~# ./event-client.rb 
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
EVENT API one.zone.raftstatus 1
^CTraceback (most recent call last):
	6: from ./event-client.rb:31:in `<main>'
	5: from ./event-client.rb:31:in `loop'
	4: from ./event-client.rb:33:in `block in <main>'
	3: from /usr/share/one/gems-dist/gems/ffi-rzmq-2.0.7/lib/ffi-rzmq/socket.rb:342:in `recv_string'
	2: from /usr/share/one/gems-dist/gems/ffi-rzmq-2.0.7/lib/ffi-rzmq/socket.rb:321:in `recvmsg'
	1: from /usr/share/one/gems-dist/gems/ffi-rzmq-2.0.7/lib/ffi-rzmq/socket.rb:584:in `__recvmsg__'
/usr/share/one/gems-dist/gems/ffi-rzmq-2.0.7/lib/ffi-rzmq/socket.rb:584:in `zmq_recvmsg': Interrupt

root@one0:~#
```
Please let me know if I can further help.

Kind regards,
Momo.

**Comment by MedicMomcilo** [comment]
Hi @rsmontero

I've attached the output of gdb with debugging symbols.

[output.txt](https://github.com/OpenNebula/one/files/4245023/output.txt)

Kind regards,
Momo.

**Comment by rsmontero** [comment]
I do not see any strange behavior on the traces. Also the string output does not give any clue about what is in the memory pages. Also it seems that you are getting events so the memory problem solved in 5.10.2 should not be affecting you.

Let's try another thing. Could you restart OpenNebula, and send us oned.log, we want to capture the log lines from start to the point oned is more than 1G. Additionally onehost list -x and onevm list -x would be needed.

You can send me the files directly if you do not want to post them here. Our plan is try to reproduce your scenario as we are not able to reproduce your issue yet.
